### DISCUSSION TITLE
Running a script on first boot

### DISCUSSION THREAD

**Discussion by dshaw619** [gitbls/sdm#74] [Q&A]
I'm trying to sort out exactly how to set up a script to run after my fully customized and burned image has been booted in an RPi (e.g. perform a full image backup to a NAS).  I think I may have it now, but thought I'd better clarify a couple of things in the docs.

https://github.com/gitbls/sdm/blob/master/Docs/Script-Details.md under 1piboot/* says "If enabled, the custom scripts in 1piboot/0*-*.sh are run when the system first boots ..." 
https://github.com/gitbls/sdm/blob/master/Docs/Command-Details.md under `--bootscripts` says "Directs sdm-firstboot to run the boot scripts in 1piboot/*.sh."  

I suspect the custom scripts should go in /usr/local/sdm/1piboot not a further subdirectory named <something>.sh.  Is that correct?  Is there any naming requirement for the scripts as might be implied by "0*-*.sh" ?

**Comment by gitbls** [comment]
Noting for clarity, these scripts run exactly once: during the very first boot of the operating system. If you want something to run every boot of the system, you have to use one of the other methods: systemd, cron, etc. 

The correct script name pattern is 0*-*.sh, so they're both wrong. Will update the docs.

You can also put scripts in /etc/sdm/0piboot/0*-*.sh. The key difference between 0piboot and 1piboot is that you can control whether the scripts in 1piboot are executed with the `--bootscripts` command. 0piboot scripts are always run during the first boot of the system.

**Reply by dshaw619** [reply]
Yes, I am interested here in running a script just once.  I do have other systemd services and cron jobs.

Looks like I somehow messed up the quote from https://github.com/gitbls/sdm/blob/master/Docs/Script-Details.md.  It does say "1piboot/0*-*.sh" so if I understand your reply correctly, that reference is correct.

I suppose it's a nit, but isn't it the case that the script named in accordance with that pattern should be put in 1piboot rather than IN "1piboot/0*-*.sh" ?  To me, that sounds like 1piboot/0*-*.sh is to be a directory.  

I read this pattern to mean that the script must be named with a leading "0", some arbitrary characters, a "-", some more arbitrary characters, followed by ".sh".  Is that correct?

Is 0piboot discussed somewhere in the docs?

**Reply by gitbls** [reply]
- 1piboot vs 0piboot. See my comment above about putting scripts in 0piboot, and the difference between 0piboot and 1piboot (Scripts in 1piboot are controlled via `--bootscripts`. Scripts in 0piboot are always run, regardless of the `--bootscripts` setting. That's the only difference. No difference in filename patterns.
- Your pattern interpretation is correct!
- 0piboot is mentioned [here (Custom-Phase-Script.md)](https://github.com/gitbls/sdm/blob/master/Docs/Custom-Phase-Script.md)
